Ensure that schema and DB changes are reported to the changes JSON file.

samueljseay opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

There was a mistake made (by me) originally when generating the changes json file for the code-analyzer. It added the DB changes to the `schema` prop and it didn't report schema changes. This resolves that ensuring that all the change data is added to the changes JSON file and under more correct prop names.

I have also as a follow up needed to refactor a few things to get this working:

1. `printSchemaChange` has the logic for reporting schema changes in it, I've adjusted it to return this information so we can store and output to the JSON file.

2. There was a bug where no `await` happened on `scanChanges`
3. There was a bug in `scanHooks`, it would exit with error if it found no description, I don't think that behaviour is desirable, we want to see all errors.

### How to test the changes in this Pull Request:

* Run code-analyzer from the `tools/code-analyzer` dir, e.g. `./bin/dev analyzer -s "https://github.com/woocommerce/woocommerce.git" "b7931409f2a553bea765f0fda421f182c58d8e4b"`
* See that the command generated a `changes.json` file containing changes with the following props: `schema`, `db`, `hooks`, `templates`

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

samueljseay committed:
Ensure that schema and DB changes are reported to the changes JSON file.

roykho commented:
Hi @samueljseay - In your reproducing steps, do you mind showing how this tool is run? An example command would be nice. Thanks!

samueljseay committed:
Fix bugs with reporting, return diff from print as a workaround for now.

samueljseay commented on tools/code-analyzer/src/commands/analyzer/index.ts:
```diff
@@ -140,9 +140,9 @@ export default class Analyzer extends Command {
 			CliUx.ux.action.stop();
 
-			this.scanChanges( diff, pluginData[ 0 ], flags, schemaDiff );
+			await this.scanChanges( diff, pluginData[ 0 ], flags, schemaDiff );
```
This was an existing bug

samueljseay commented on tools/code-analyzer/src/commands/analyzer/index.ts:
```diff
@@ -233,15 +233,11 @@ export default class Analyzer extends Command {
-			hooks: Object.fromEntries( hooks.entries() ),
-			schema: databaseUpdates || {},
-		} );
+		let schemaDiffResult = {};
```
I wouldn't normally declare a let variable like this for collecting the schema change, but I had to do some refactoring of `printSchemaChange` to get the data I needed. I think `code-analyzer` needs a bit more work to be more generally useful and I've logged an issue for this on the Atlas team board.

samueljseay commented on tools/code-analyzer/src/commands/analyzer/index.ts:
```diff
@@ -445,11 +448,12 @@ export default class Analyzer extends Command {
 				const name = getHookName( hookName[ 3 ] );
 
-				const description = getHookDescription( raw, name );
+				const description = getHookDescription( raw, name ) || '';
```
This was upsetting TS since `getHookDescription` returns `string | false`, we coerce to empty string if its false here to stop TS complaining on line https://github.com/woocommerce/woocommerce/pull/34238/files#diff-d71f73de77bb524fe98ceab4c4649eb82cb161cbf1a070d4f463c321c76b267cL470

samueljseay commented on tools/code-analyzer/src/print.ts:
```diff
@@ -114,9 +114,10 @@ export const printSchemaChange = (
 	log: ( s: string ) => void
-): void => {
+): Record< string, string > => {
+	const diff: Record< string, string > = {};
```
This is a quick workaround to get the info I need from this function. Refactoring in future would be needed and I've raised an issue about it.

samueljseay commented:
Apologies for the lack of detail @roykho I've updated testing instructions with a sample command.

Your request prompted me to find some other issues too so I'm glad for that, I've updated the description for that.

roykho reviewed (approved):
Tested well. This is probably unrelated but I am seeing some errors such as this `Error: Hook woocommerce_max_terms_displayed_in_attributes_page has no description. Please add a description.` which claims descriptions were not added but when I checked the code, the description is there.

But otherwise LGTM. Good job!

samueljseay commented:
@roykho I have seen that issue too, I think it was being hidden by some of the bugs present here. I will explore a fix for why this is happening.
